File: examples/normal_form/channel_map.py

import numpy as np
import synergia as syn

def run():

    # logger
    screen = syn.utils.parallel_utils.Logger(0, 
            syn.utils.parallel_utils.LoggerV.DEBUG)

    simlog = syn.utils.parallel_utils.Logger(0, 
            syn.utils.parallel_utils.LoggerV.INFO_TURN)

    # lattice
    reader = syn.lattice.MadX_reader()
    lattice = reader.get_lattice("fodo", "channel.madx")
    syn.simulation.Lattice_simulator.tune_circular_lattice(lattice)

    # calculate one turn map at given order (3 here)
    mapping = syn.simulation.Lattice_simulator.get_one_turn_map_o3(lattice)

    # save
    mapping.save_json("mapping.json")

    # load
    m2 = syn.foundation.TMapping_o3.load_json("mapping.json")

    # create the normal form object from one turn mapping directly
    ref = lattice.get_reference_particle()
    e0 = ref.get_total_energy()
    pc0 = ref.get_momentum()
    mass = ref.get_mass()

    nf = syn.foundation.NormalForm_o3(m2, e0, pc0, mass)

    # save nf
    nf.save_json("nf.json")

    # load
    nf2 = syn.foundation.NormalForm_o3.load_json("nf.json")

    # convert the mapping to a json object
    # this currently fails
    # TypeError: Unregistered type : nlohmann::json_abi_v3_11_3::basic_json<std::map, std::vector, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, long, unsigned long, double, std::allocator, nlohmann::json_abi_v3_11_3::adl_serializer, std::vector<unsigned char, std::allocator<unsigned char> >, void>
    # mapping_json = mapping.to_json()
    # print(mapping_json)

    # or iterate through the components and fields
    for comp in range(6):
        trigon = mapping.component(comp)
        print("\n\nComponent {}".format(comp))

        for pwr in range(trigon.power()+1):
           for idx in range(trigon.count(pwr)): 
               idx_to_exp = "syn.foundation.Trigon_index_to_exp_o{}(idx)".format(pwr)

               print('power = {}, exp = {}, term = {}'.format( 
                    pwr, eval(idx_to_exp), trigon.get_term(pwr, idx)))



print("channel_map")
run()

File: examples/normal_form/channel_map.cc

#include "synergia/lattice/madx_reader.h"
#include "synergia/simulation/independent_stepper_elements.h"
#include "synergia/simulation/lattice_simulator.h"
#include "synergia/simulation/propagator.h"
#include "synergia/utils/cereal_files.h"
#include "synergia/utils/utils.h"

#include <iostream>

void
run()
{
  Logger screen(0, LoggerV::DEBUG);
  Logger log(0, LoggerV::INFO_TURN);

  // read lattice
  MadX_reader reader;
  auto lattice = reader.get_lattice("model", "foborodobo128.madx");

  // tune the lattice
  lattice.set_all_string_attribute("extractor_type", "libff");
  Lattice_simulator::tune_circular_lattice(lattice);

  lattice.print(screen);

  // one turn map
  const int order =
    3; // higher orders use more memory and may not compile on many systems
  auto mapping = Lattice_simulator::get_one_turn_map<order>(lattice);
  // std::cout << mapping.to_json() << "\n";

  // save
  json_save(mapping, "mapping.json");

  // load
  TMapping<Trigon<double, order, 6>> m2;
  json_load(m2, "mapping.json");

  // ref
  auto const& ref = lattice.get_reference_particle();
  auto energy = ref.get_total_energy();
  auto momentum = ref.get_momentum();
  auto mass = ref.get_mass();

  // construct the normal from from loaded mapping
  auto nf = NormalForm<order>(mapping, energy, momentum, mass);

  // save the nf object
  json_save(nf, "nf.json");

  // load
  NormalForm<order> nf2;
  json_load(nf2, "nf.json");

  // print something so it doesnt get optimized away by the compiler
  std::cout << "nf.f[0] = " << nf.get_f()[1].to_json() << "\n";
  std::cout << "nf2.f[0] = " << nf2.get_f()[1].to_json() << "\n";

  for (int comp = 0; comp < 6; ++comp) {
    auto trigon = mapping[comp];
    std::cout << "\n\nComponent " << comp << "\n";

    trigon.each_term([&trigon](int i, auto const& inds, auto term) {
      std::cout << "power = " << inds.size();

      std::array<int, trigon.dim> exp{0};
      for (auto const& x : inds) ++exp[x];

      std::cout << ", exp = [";
      for (int x = 0; x < exp.size() - 1; ++x) std::cout << exp[x] << ", ";
      std::cout << exp.back() << "]";

      std::cout << ", term = " << term << "\n";
    });
  }
}

int
main(int argc, char** argv)
{
  synergia::initialize(argc, argv);

  run();

  synergia::finalize();
  return 0;
}
